### PR TITLE
[rebass] Support standard HTML attributes as Rebass component props

### PR DESCRIPTION
**Pull request by chuckdries**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://rebassjs.org/Button  
> note: Rebass is a package that simply wraps native HTML elements. This documentation link is to demonstrate that the Button component is indeed a native button and should behave as such. The previous type definitions simply hard-coded a select set of common HTML attributes instead of adding them properly. 
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by chuckdries** [comment]
I'm a bit of a typescript novice still - can whoever reviews this explain to me why the types for all those components needed to be union types? I initially tried intersection types, which made more conceptual sense to me, but it caused the tests to fail

**Comment by chuckdries** [comment]
Wait these should be intersection types - I'll get to them in the morning

**Comment by armanio123** [comment]
@chuckdries The correct syntax should use the intersection type (`&`) as you want to combine both interfaces.

You are getting an error because there's a conflict with a couple of properties (e.g. `width`) making it pretty much unusable.

Let's take a deeper look at the `BoxProps.width` property. Your `d.ts` file defines as `number | string | ReadonlyArray<number>` and React defines it as `number | string`. The result of intersecting that would be `number | string | (ReadonlyArray<number> & string) | (ReadonlyArray<number> & number)`. So, assigning `width={[1,2,3,4]}` will not work as it doesn't match the type.

To fix this you need to specify which type is the correct one by "omiting" the properties you don't want (in this case React's). You need to define an `Omit` type that will help with that. Let's use `Card` as an example:

```ts
// Pick all properties from T excluding the ones in K
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

interface CardKnownProps extends BoxProps {
    border?: number | string;
    borderColor?: string;
    borderRadius?: number | string;
    boxShadow?: string;
    backgroundImage?: string;
    backgroundSize?: string;
    backgroundPosition?: string;
    backgroundRepeat?: string;
    opacity?: number;
    variant?: string;
}

// This could have been an intersection but I prefer interface as it's easier to read and also more efficient for tooling.
export interface CardProps extends CardKnownProps, Omit<React.HTMLProps<HTMLDivElement>, keyof CardKnownProps> {}

export const Card: React.FunctionComponent<CardProps>;
```

Hope this makes sense. You then can use `Omit` for all of your exports to avoid this error.

File: types/rebass/index.d.ts [hunk 1]
```diff
@@ -63,28 +66,24 @@ export interface CardProps extends BoxProps {
...
-    alt?: string;
 }
-export const Image: React.FunctionComponent<ImageProps>;
+export const Image: React.FunctionComponent<ImageProps | React.HTMLProps<HTMLImageElement>>;
 
-export interface LinkProps extends BoxProps {
-    href?: string;
-}
-export const Link: React.FunctionComponent<LinkProps>;
+// export interface LinkProps extends BoxProps {}
```
Comment (armanio123) [inline types/rebass/index.d.ts]: Remove unused comments.

**Review by armanio123 — CHANGES_REQUESTED** [review]

**Comment by chuckdries** [comment]
Closing in favor of https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32930, which incorporates this work but adds more accurate types for the style props. Thanks @angusfretwell!

**Pull request closed without merging**